allennlp/modules/seq2seq_encoders/stacked_self_attention.py

```diff
@@ -72,37 +71,33 @@ def __init__(self,
         super(StackedSelfAttentionEncoder, self).__init__()
 
         self._use_positional_encoding = use_positional_encoding
-        self._attention_layers: List[MultiHeadSelfAttention] = []
-        self._feedfoward_layers: List[FeedForward] = []
-        self._layer_norm_layers: List[LayerNorm] = []
-        self._feed_forward_layer_norm_layers: List[LayerNorm] = []
+        self._attention_layers = ModuleList([])
+        self._feedfoward_layers = ModuleList([])
+        self._layer_norm_layers = ModuleList([])
+        self._feed_forward_layer_norm_layers = ModuleList([])
 
         feedfoward_input_dim = input_dim
-        for i in range(num_layers):
+        for _ in range(num_layers):
             feedfoward = FeedForward(feedfoward_input_dim,
                                      activations=[Activation.by_name('relu')(),
                                                   Activation.by_name('linear')()],
                                      hidden_dims=[feedforward_hidden_dim, hidden_dim],
                                      num_layers=2,
                                      dropout=dropout_prob)
 
-            self.add_module(f"feedforward_{i}", feedfoward)
```

Comment:
@matt-gardner, Ah, I had forgotten that `ModuleList` would change the names. Somewhat surprising that `add_module` isn't sufficient. Unfortunate. However, I think we can work around this and preserve backwards compatibility by using, e.g.:

`setattr(self, f"feedforward_{i}", feedforward)`

@JizeCao, could you please give that a try?

Reply:
Pytorch does crazy stuff with `setattr`; I don't think this is a good idea.  It would make interactions with pytorch magic very hard to reason about.

Reply:
@matt-gardner , I'm not sure I understand your concern. We use Pytorch's `__setattr__` every single time we modify any instance variable of a `Module`. The only difference here is that the name is dynamic. Example:

```
In [1]: class Foo:
   ...:     def __setattr__(self, name, value):
   ...:         print("FOO")
   ...:         object.__setattr__(self, name, value)
   ...:

In [2]: f = Foo()

In [3]: f.x = 7
FOO
```

Reply:
Are you suggesting adding a call to `setattr` in addition to appending to the module list?  That's what I was expecting.  Having two calls will definitely break the pytorch module, because pytorch will list the parameters twice (it registers stuff at every call to `setattr`).  I think this is a case where we should just break backwards compatibility, because there's a workaround and because it's fixing a bug.

Reply:
No, I'm suggesting using `setattr(self, f"feedforward_{i}", feedforward)` alone. This is very similar to the original code. Its dispatches to the exact same `__setattr__` that any assignment to `self` does. Preserving backwards compatibility while fixing the bug seems quite simple here.

Reply:
I'd be very surprised if that actually worked, because you'd expect pytorch to have the same logic there as it does when you call `add_module`, which we know doesn't work.  But, having a test would tell us one way or the other.

Reply:
another solution would be to override `load_state_dict` in the following way:

```python

@overrides
def load_state_dict(self, state_dict, strict=True):
    new_state = {}
    for name, param in state_dict.items():
        # Depreciation warning here
        if name == ....
            name = new_name
        new_state[name] = param
    super().load_state_dict(new_state, strict=strict)

```

I think that would work?

Reply:
I dug around in pytorch's [load_state_dict](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py) code, I don't think it's recursively implemented.  It was hard to tell, so I could be wrong, but I don't _think_ this would actually work.

Reply:
it is recursively implemented:
https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L759

Reply:
Except that's recursing on this private method, _not_ `load_state_dict()` - I don't think it would do anything unless you overrode the private method.  That's what I was confused about.  But, yeah, if you do it in the private method it'd probably work.